Make `POSIXSignalTest.siq_queue_EPERM` consider elevated permissions

Krebrov001 opened the pull request:
Add a function `has_elevated_permissions()` to check if the current user
has elevated permissions by examining the `CapEff` field in the status
file of the running process's proc entry.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

- Relates to #2382

Krebrov001 committed:
Make `POSIXSignalTest.siq_queue_EPERM` consider elevated permissions

Add a function `has_elevated_permissions()` to check if the current user
has elevated permissions by examining the `CapEff` field in the status
file of the running process's proc entry.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 commented on service/test/POSIXSignalTest.cpp:
```diff
@@ -61,6 +69,40 @@ void POSIXSignalTest::TearDown(void)
+bool POSIXSignalTest::has_elevated_permissions(void) const
+{
+    uint64_t cap = 0;
+    uint64_t cap_sys_admin = 0x00200000;  // TODO: magic number, is there a #define for it?
```
magic number, is there a `#define` for it?

cmcantalupo replied:
We probably just want to check for CAP_KILL which is 0x20, but perhaps renaming the function as well "has_cap_kill()"

Krebrov001 commented on service/test/POSIXSignalTest.cpp:
```diff
@@ -201,10 +243,13 @@ TEST_F(POSIXSignalTest, sig_queue_ESRCH)
     }
-    else {
+    else if (has_elevated_permissions()) {  // the non root user with elevated permissions
+        m_posix_sig->sig_queue(1, SIGCONT, 2);  // TODO: how to do EXPECT_NO_THROW?
```
How to do `EXPECT_NO_THROW`? Like for any other non root user, we used a `GEOPM_EXPECT_THROW_MESSAGE` to say that we are expecting `sig_queue()` to necessarily fail. So if it doesn't throw an exception, the test would fail.

For the non root user with elevated permissions, can we have a similar assertion that excepts the call `sig_queue()` to necessarily **not** fail, or necessarily **not** throw an exception. So if it throws an exception, the test would fail, and if it doesn't throw, the test would pass. How can we do that?

cmcantalupo replied:
If the test throws an unhandled exception, gtest considers it a failure, that is why there isn't an "EXPECT_NO_THROW" macro.  Everything is expected not to throw :)

Krebrov001 replied:
> If the test throws an unhandled exception, gtest considers it a failure, that is why there isn't an "EXPECT_NO_THROW" macro. Everything is expected not to throw :)

So nothing else needs to be done here?

Krebrov001 reviewed (commented):
Please answer my questions.

cmcantalupo commented on service/test/POSIXSignalTest.cpp:
```diff
@@ -61,6 +69,40 @@ void POSIXSignalTest::TearDown(void)
+    std::string status_path = "/proc/" + std::to_string(pid) + "/status";
+
+    // TODO: This is a very memory inefficient algorithm and should be replaced some day.
+    try {
```
Why the catch all?  Under what conditions would we expect this function to throw?

cmcantalupo reviewed (changes requested):
Looks pretty good.  Just a couple of comments.

Krebrov001 committed:
Addressing review comments

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Refactor `geopm::read_file()` in `Helper.cpp`

Replace the old implementation with a newer, more robust one, using
stringstreams. The reason why was because the old implementation was
unable to correctly open files in the procfs.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

cmcantalupo reviewed (approved):
LGTM
